File: Foreign/SonyLE/XLELayer/PlacementManipulator.cpp
// Copyright 2015 XLGAMES Inc.
//
// Distributed under the MIT License (See
// accompanying file "LICENSE" or the website
// http://www.opensource.org/licenses/mit-license.php)

#pragma warning(disable:4564)

#include "NativeManipulators.h"
#include "XLELayerUtils.h"

using namespace System::ComponentModel;
using namespace System::ComponentModel::Composition;
using namespace System::Windows::Forms;

namespace XLELayer
{
    public interface class IXLEAssetService
    {
    public:
        virtual System::String^ AsAssetName(System::Uri^ uri) = 0;
        virtual String^ StripExtension(String^ input) = 0;
        virtual String^ GetBaseTextureName(String^ input) = 0;
    };

    [Export(IXLEAssetService::typeid)]
    [PartCreationPolicy(CreationPolicy::Shared)]
    public ref class XLEAssetService : public IXLEAssetService
    {
    public:
        virtual System::String^ AsAssetName(System::Uri^ uri)
        {
                // covert this uri into a string filename that is fit for the assets system
            if (uri->IsAbsoluteUri) {
                auto cwd = gcnew Uri(System::IO::Directory::GetCurrentDirectory()->TrimEnd('\\') + "\\");
                auto relUri = cwd->MakeRelativeUri(uri);
                return Uri::UnescapeDataString(relUri->OriginalString);
            } else {
                return Uri::UnescapeDataString(uri->OriginalString);
            }
        }

        virtual String^ StripExtension(String^ input)
        {
            int dot = input->LastIndexOf('.');
            int sep0 = input->LastIndexOf('/');
            int sep1 = input->LastIndexOf('\\');
            if (dot > 0 && dot > sep0 && dot > sep1) {
                return input->Substring(0, dot);
            }
            return input;
        }

        virtual String^ GetBaseTextureName(String^ input)
        {
                // to get the "base texture name", we must strip off _ddn, _df and _sp suffixes
            auto withoutExt = StripExtension(input);
            if (withoutExt->EndsWith("_ddn", true, System::Globalization::CultureInfo::CurrentCulture)) {
                return withoutExt->Substring(0, withoutExt->Length-4);
            }
            if (    withoutExt->EndsWith("_df", true, System::Globalization::CultureInfo::CurrentCulture)
                ||  withoutExt->EndsWith("_sp", true, System::Globalization::CultureInfo::CurrentCulture)) {
                return withoutExt->Substring(0, withoutExt->Length-3);
            }
            return withoutExt;
        }
    };

    public interface class IPlacementControls
    {
    public:
        property ActiveManipulatorContext^ ActiveContext 
            { virtual void set(ActiveManipulatorContext^); }
    };

    [Export(LevelEditorCore::IManipulator::typeid)]
    [Export(IInitializable::typeid)]
    [PartCreationPolicy(CreationPolicy::Shared)]
    public ref class PlacementManipulator : public LevelEditorCore::IManipulator, public IInitializable
    {
    public:
        virtual bool Pick(LevelEditorCore::ViewControl^ vc, Point scrPt)        { return _nativeManip->MouseMove(vc, scrPt); }
        virtual void Render(LevelEditorCore::ViewControl^ vc)                   { _nativeManip->Render(vc); }
        virtual void OnBeginDrag()                                              { _nativeManip->OnBeginDrag(); }
        virtual void OnDragging(LevelEditorCore::ViewControl^ vc, Point scrPt)  { _nativeManip->OnDragging(vc, scrPt); }

        virtual void OnEndDrag(LevelEditorCore::ViewControl^ vc, Point scrPt) 
		{
            _nativeManip->OnEndDrag(vc, scrPt);
		}

        property LevelEditorCore::ManipulatorInfo^ ManipulatorInfo
        {
            virtual LevelEditorCore::ManipulatorInfo^ get() 
            {
                return gcnew LevelEditorCore::ManipulatorInfo(
                    Sce::Atf::Localizer::Localize("Placements", String::Empty),
                    Sce::Atf::Localizer::Localize("Activate placements editing", String::Empty),
                    LevelEditorCore::Resources::TerrainManipImage,
                    Keys::None);
            }
        }

        virtual void Initialize()
        {
            _manipSettings = gcnew Settings();
            _manipContext = gcnew ActiveManipulatorContext();
            _manipContext->ManipulatorSet = NativeManipulatorLayer::SceneManager->CreatePlacementManipulators(_manipSettings);
            _controls->ActiveContext = _manipContext;
            _nativeManip = gcnew NativeManipulatorLayer(_manipContext);

            if (_resourceLister) {
                _resourceLister->SelectionChanged += 
                    gcnew EventHandler(this, &PlacementManipulator::resourceLister_SelectionChanged);
            }
        }

    private:
        ref class Settings : GUILayer::IPlacementManipulatorSettingsLayer
        {
        public:
            virtual String^ GetSelectedModel() override { return _selectedModel; }
            virtual void SelectModel(String^ newModelName) override  { _selectedModel = newModelName; }
            virtual void EnableSelectedModelDisplay(bool newState) override {}
            virtual void SwitchToMode(unsigned newMode) override {}

            Settings()
            {
                _selectedModel = "game/model/nature/bushtree/BushE";
            }

            String^ _selectedModel;
        };

        NativeManipulatorLayer^ _nativeManip;
        ActiveManipulatorContext^ _manipContext;
        Settings^ _manipSettings;

        [Import(AllowDefault = false)] IPlacementControls^ _controls;
        [Import(AllowDefault =  true)] LevelEditorCore::ResourceLister^ _resourceLister;
        [Import(AllowDefault = false)] IXLEAssetService^ _assetService;

        void resourceLister_SelectionChanged(Object^ sender, EventArgs^ e)
        {
            auto resourceUri = _resourceLister->LastSelected;
            if (resourceUri) {
                _manipSettings->_selectedModel = 
                    _assetService->StripExtension(_assetService->AsAssetName(resourceUri));
            }
        }
    };
}

File: Foreign/SonyLE/LevelEditorXLE/Placements/PlacementManipulator.cs
// Copyright 2015 XLGAMES Inc.
//
// Distributed under the MIT License (See
// accompanying file "LICENSE" or the website
// http://www.opensource.org/licenses/mit-license.php)

using System;
using System.ComponentModel;
using System.ComponentModel.Composition;
using System.Windows.Forms;
using System.Drawing;
using Sce.Atf;
using Sce.Atf.Applications;
using LevelEditorXLE.Extensions;
using LevelEditorCore;

#pragma warning disable 0649 // Field '...' is never assigned to, and will always have its default value null

namespace LevelEditorXLE.Placements
{
    [Export(typeof(LevelEditorCore.IManipulator))]
    [Export(typeof(XLEBridgeUtils.IShutdownWithEngine))]
    [PartCreationPolicy(CreationPolicy.Shared)]
    public class PlacementManipulator : IManipulator, XLEBridgeUtils.IShutdownWithEngine, IDisposable
    {
        public ManipulatorPickResult Pick(ViewControl vc, Point scrPt)
        {
            if (_nativeManip.MouseMove(vc as GUILayer.IViewContext, scrPt))
                return ManipulatorPickResult.ImmediateBeginDrag;
            return ManipulatorPickResult.Miss;
        }
        public void OnBeginDrag(ViewControl vc, Point scrPt)   { _nativeManip.OnBeginDrag(vc as GUILayer.IViewContext, scrPt); }
        public void OnDragging(ViewControl vc, Point scrPt)    { _nativeManip.OnDragging(vc as GUILayer.IViewContext, scrPt); }
        public void OnEndDrag(ViewControl vc, Point scrPt)     { _nativeManip.OnEndDrag(vc as GUILayer.IViewContext, scrPt); }
        public void OnMouseWheel(ViewControl vc, Point scrPt, int delta) { _nativeManip.OnMouseWheel(vc as GUILayer.IViewContext, scrPt, delta); }

        public void Render(object opaqueContext, ViewControl vc) 
        {
            var context = opaqueContext as GUILayer.SimpleRenderingContext;
            if (context == null) return;
            _nativeManip.Render(context); 
        }

        public ManipulatorInfo ManipulatorInfo
        {
            get
            {
                return new ManipulatorInfo(
                    "Placements".Localize(),
                    "Activate placements editing".Localize(),
                    LevelEditorCore.Resources.TerrainManipImage,
                    Keys.None);
            }
        }

        public void Shutdown()
        {
            var r = _contextRegistry.Target as IContextRegistry;
            if (r != null) r.ActiveContextChanged -= OnActiveContextChanged;

            var rl = _resourceLister.Target as LevelEditorCore.ResourceLister;
            if (rl != null) rl.SelectionChanged -= resourceLister_SelectionChanged;

            if (_nativeManip != null) { _nativeManip.Dispose(); _nativeManip = null; }
            if (_manipContext != null) { _manipContext.Dispose(); _manipContext = null; }
        }

        [ImportingConstructor]
        public PlacementManipulator(IContextRegistry contextRegistry, LevelEditorCore.ResourceLister resourceLister)
        {
            _manipContext = new GUILayer.ActiveManipulatorContext();
            _nativeManip = null;

            _manipSettings = new Settings();
            _nativeManip = new GUILayer.NativeManipulatorLayer(_manipContext);

            if (resourceLister != null)
            {
                resourceLister.SelectionChanged += resourceLister_SelectionChanged;
                _resourceLister = new WeakReference(resourceLister);
            }

            _contextRegistry = new WeakReference(contextRegistry);
            contextRegistry.ActiveContextChanged += OnActiveContextChanged;
            _nativeManip = new GUILayer.NativeManipulatorLayer(_manipContext);
        }

        public void Dispose() { Shutdown(); }

        private class Settings : GUILayer.IPlacementManipulatorSettingsLayer
        {
            public override string GetSelectedModel() { return _selectedModel; }
            public override string GetSelectedMaterial() { return _selectedMaterial; }
            public override void SelectModel(string newModelName, string materialName) { _selectedModel = newModelName; _selectedMaterial = materialName; }
            public override void EnableSelectedModelDisplay(bool newState) {}
            public override void SwitchToMode(uint newMode) {}

            internal Settings()
            {
                _selectedModel = "game/model/galleon/galleon";
                _selectedMaterial = "game/model/galleon/galleon";
            }

            internal String _selectedModel;
            internal String _selectedMaterial;
        };

        private void OnActiveContextChanged(object obj, EventArgs args)
        {
            GUILayer.EditorSceneManager sceneMan = null;

            IContextRegistry registry = obj as IContextRegistry;
            if (registry != null)
            {
                var gameExt = registry.GetActiveContext<Game.GameExtensions>();
                if (gameExt != null)
                    sceneMan = gameExt.SceneManager;
            }

            if (sceneMan != null)
            {
                _manipContext.ManipulatorSet = sceneMan.CreatePlacementManipulators(_manipSettings);
            }
            else
            {
                _manipContext.ManipulatorSet = null;
            }
        }

        public System.Windows.Forms.Control GetHoveringControl() { return null; }
        public event System.EventHandler OnHoveringControlChanged;

        private GUILayer.NativeManipulatorLayer _nativeManip;
        private GUILayer.ActiveManipulatorContext _manipContext;
        private Settings _manipSettings;

        [Import(AllowDefault = false)] IXLEAssetService _assetService;
        WeakReference _contextRegistry;
        WeakReference _resourceLister;

        void resourceLister_SelectionChanged(object sender, EventArgs e)
        {
            var resLister = sender as LevelEditorCore.ResourceLister;
            var resourceUri = resLister.LastSelected;
            if (resourceUri != null) {
                _manipSettings._selectedModel = 
                    _assetService.StripExtension(_assetService.AsAssetName(resourceUri));
                _manipSettings._selectedMaterial = _manipSettings._selectedModel;
            }
        }
    };
}

